project/src/demo/puzzle/level/level-rank-demo-old.gd

```diff
@@ -100,29 +101,31 @@ func _calculate_combo_factor() -> void:
 	var combo_factor_min := 0.0
 	var combo_factor_max := 4.0
 	for _i in range(20):
-		CurrentLevel.settings.rank.combo_factor = \
+		CurrentLevel.settings.rank.legacy_rules["combo_factor"] = \
 				0.5 * (combo_factor_min + combo_factor_max)
-		var combo_score_for_grade := CurrentLevel.settings.rank.combo_factor \
-				* RankCalculator.MASTER_COMBO_SCORE * pow(RankCalculator.RDF_COMBO_SCORE_PER_LINE, target_rank)
-		if combo_score_for_grade >= best_result.combo_score_per_line:
-			combo_factor_max = CurrentLevel.settings.rank.combo_factor
+		var combo_score_for_grade := \
+				float(CurrentLevel.settings.rank.legacy_rules.get("combo_factor", 1.0)) \
+				* LegacyRankCalculator.MASTER_COMBO_SCORE \
+				* pow(LegacyRankCalculator.RDF_COMBO_SCORE_PER_LINE, target_rank)
+		if combo_score_for_grade >= best_result.combo_score / float(best_result.lines):
+			combo_factor_max = CurrentLevel.settings.rank.legacy_rules.get("combo_factor", 1.0)
 		else:
-			combo_factor_min = CurrentLevel.settings.rank.combo_factor
+			combo_factor_min = CurrentLevel.settings.rank.legacy_rules.get("combo_factor", 1.0)
 	
-	_text_edit.text += "\"combo_factor %.2f\",\n" % [CurrentLevel.settings.rank.combo_factor]
+	_text_edit.text += "\"combo_factor %.2f\",\n" % [CurrentLevel.settings.rank.legacy_rules.get("combo_factor", 1.0)]
 
 
 ## Calculates and outputs the master_pickup_score_per_line for levels with pickups.
 func _calculate_master_pickup_score_per_line() -> void:
 	var target_rank := _target_rank()
 	var best_result := _best_result()
 	
-	CurrentLevel.settings.rank.master_pickup_score_per_line = \
-			best_result.pickup_score_per_line \
-			/ pow(RankCalculator.RDF_PICKUP_SCORE_PER_LINE, target_rank)
+	CurrentLevel.settings.rank.legacy_rules["master_pickup_score_per_line"] = \
+			(best_result.pickup_score / float(best_result.lines)) \
```

Comment:
is there a risk of division by zero?

Reply:
Yes, the old code specifically prevented this with 'max(lines, 1)'

Reply:
Changed division to `(best_result.pickup_score / max(best_result.lines, 1)) \` to avoid division by zero